### PR TITLE
Restructure cpp code

### PR DESCRIPTION
**Pull request by mscroggs**
- Move functions from of `finite-element` to `finite-element-utils`, so `finite-element` now only contains the class `FiniteElement`
- Added `basix.h`/`basix.cpp` that includes all of basix so C++ users can import just one file

Coupled with https://github.com/FEniCS/dolfinx/pull/3191

### CODE REVIEW

File: cpp/basix/basix.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+namespace basix
+{
+
+/// Return the Basix version number
+/// @return version string
+std::string version();
+
+// using basix::create_custom_element;
+// using basix::create_element;
+// using basix::FiniteElement;
```
Comment (mscroggs) [inline cpp/basix/basix.h:26]:
```suggestion

```

File: cpp/basix/finite-element-utils.cpp [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+#include "e-regge.h"
+#include "e-serendipity.h"
+#include "math.h"
+#include "polyset.h"
+#include <basix/version.h>
+#include <cmath>
+#include <concepts>
+#include <limits>
+#include <numeric>
+#define str_macro(X) #X
```
Comment (garth-wells) [inline cpp/basix/finite-element-utils.cpp:25]: Add a space here

**Review by garth-wells — COMMENTED** [review]
The name `finite-element-utils.h` is a bit clumsy. Could it be shorter?

File: cpp/basix/e-brezzi-douglas-marini.cpp [hunk 3]
```diff
@@ -6,6 +6,7 @@
 #include "e-lagrange.h"
 #include "e-nedelec.h"
 #include "element-families.h"
+#include "finite-element-utils.h"
```
Comment (garth-wells) [inline cpp/basix/e-brezzi-douglas-marini.cpp:9]: Looks like `finite-element.h` is missing - the `FiniteElement` constructor is used in this file.

File: cpp/basix/finite-element-utils.cpp [hunk 4]
```diff
@@ -0,0 +1,668 @@
...
+#define str(X) str_macro(X)
+
+using namespace basix;
+
+template <typename T, std::size_t d>
```
Comment (garth-wells) [inline cpp/basix/finite-element-utils.cpp:30]: Maybe use 
```cpp
template <typename T, std::size_t d>
using mdspan_t = MDSPAN_IMPL_STANDARD_NAMESPACE::mdspan<
    T, MDSPAN_IMPL_STANDARD_NAMESPACE::dextents<std::size_t, d>>;
```
here so this code won't be affected by changes to impl::mdspan_t (not even sure where this is defined). Also put inside and anonymous namespace.

File: cpp/basix/finite-element-utils.cpp [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+    xshapes[i] = std::vector(x[i].size(), std::array<std::size_t, 2>{0, tdim});
+    x_data[i].resize(x[i].size());
+    Mshapes[i] = std::vector(
+        M[i].size(), std::array<std::size_t, 4>{0, value_size, 0, nderivs});
+    M_data[i].resize(M[i].size());
+  }
+
+  std::array<std::size_t, 2> xshape = {npoints, tdim};
+  std::vector<T> xb(xshape[0] * xshape[1]);
+  MDSPAN_IMPL_STANDARD_NAMESPACE::mdspan<
```
Comment (garth-wells) [inline cpp/basix/finite-element-utils.cpp:73]: Use the shortcuts for `mdspan_t`.

File: cpp/basix/finite-element-utils.cpp [hunk 6]
```diff
@@ -0,0 +1,668 @@
...
+template <std::floating_point T>
+std::pair<std::vector<T>, std::array<std::size_t, 2>>
+element::compute_dual_matrix(
+    cell::type cell_type, polyset::type poly_type, mdspan_t<const T, 2> B,
+    const std::array<std::vector<impl::mdspan_t<const T, 2>>, 4>& x,
+    const std::array<std::vector<impl::mdspan_t<const T, 4>>, 4>& M, int degree,
+    int nderivs)
+{
+  std::size_t num_dofs(0), vs(0);
+  for (auto& Md : M)
```
Comment (garth-wells) [inline cpp/basix/finite-element-utils.cpp:119]: Is this just a shape test? Maybe add a comment on what it's doing.

File: cpp/basix/finite-element-utils.cpp [hunk 7]
```diff
@@ -0,0 +1,668 @@
...
+          for (std::size_t j = 0; j < Pt.extent(1); ++j)
+            Pt(i, j) = P(0, j, i);
+
+        std::vector<T> De_b(Me.extent(0) * Me.extent(1) * Pt.extent(1));
+        mdspan_t<T, 2> De(De_b.data(), Me.extent(0) * Me.extent(1),
+                          Pt.extent(1));
+        math::dot(mdspan_t<const T, 2>(Me.data_handle(),
+                                       Me.extent(0) * Me.extent(1),
+                                       Me.extent(2)),
+                  Pt, De);
```
Comment (garth-wells) [inline cpp/basix/finite-element-utils.cpp:186]: Formatting

File: cpp/basix/finite-element-utils.cpp [hunk 8]
```diff
@@ -0,0 +1,668 @@
...
+    {
+    case element::family::P:
+    {
+      FiniteElement<T> sub_element
+          = create_element<T>(element::family::P, cell::type::interval, degree,
+                              lvariant, dvariant, true);
+      switch (cell)
+      {
+      case cell::type::quadrilateral:
+      {
```
Comment (garth-wells) [inline cpp/basix/finite-element-utils.cpp:405]: Remove braces.

File: cpp/basix/finite-element-utils.cpp [hunk 9]
```diff
@@ -0,0 +1,668 @@
...
+            for (int k = 0; k < n; ++k)
+              perm.push_back(8 + 12 * n + 6 * n * n + i + n * j + n * n * k);
+          }
+        }
+      }
+      assert((int)perm.size() == (degree + 1) * (degree + 1) * (degree + 1));
+      break;
+    }
+    default:
+    {
```
Comment (garth-wells) [inline cpp/basix/finite-element-utils.cpp:531]: This is hard to read.

**Review by garth-wells — COMMENTED** [review]
First pass of comments.

**Pull request closed without merging**